src/type_checker/checker.rs

```diff
@@ -141,7 +141,20 @@ impl<B: Backend> TypeChecker<B> {
 
                 // check if generic is allowed
                 if fn_sig.require_monomorphization() && typed_fn_env.is_in_forloop() {
-                    return Err(self.error(ErrorKind::GenericInForLoop, expr.span));
+                    for (observed_arg, expected_arg) in args.iter().zip(fn_sig.arguments.iter()) {
+                        // check if the arg involves generic vars
+                        if !expected_arg.extract_generic_names().is_empty() {
+                            let mut forbidden_env = typed_fn_env.clone();
+                            forbidden_env.forbid_forloop_scope();
+
+                            // rewalk the observed arg expression
+                            // it should throw an error if the arg contains generic vars relating to the variables in the forloop scope
+                            self.compute_type(observed_arg, &mut forbidden_env)?;
```

Comment:
I think there is a small bug in the handling of the for loop variables. Take this code (just modifying one of the examples)
```rust
fn clone(const LEN: Field, val: Field) -> [Field; LEN] {
    let mut arr = [0; LEN];
    for idx in 0..LEN {
        arr[idx] = val;
    }
    return arr;
}

fn main(pub xx: Field) {
    let mut arr3 = [[0; 2]; 2];
    for idx in 0..2 {
        let cloned = clone(2, xx + idx);
        arr3[idx] = cloned;
    }

    let fst = arr3[0];
    let snd = arr3[1];
    assert_eq(fst[0], 0);
    assert_eq(fst[1], 0);
    assert_eq(snd[0], 1);
    assert_eq(snd[1], 1);
}
```

I would expect that adding a nested loop of one iteration inside the one defined do not change any of the emitted constraints, and indeed it works.
```rust
    for idx in 0..2 {
        for unused in 0..1 {
            let cloned = clone(2, xx + idx);
            arr3[idx] = cloned;
        }
    }
```

However, if the `LEN` argument is replaced with the outer loop variable
```rust
    for idx in 0..2 {
        for unused in 0..1 {
            let cloned = clone(idx, xx + idx);
            arr3[idx] = cloned;
        }
    }
```

then the nice error about the call of generic function is not returned anymore and instead there is this panic
```
thread 'main' panicked at src/parser/types.rs:800:14:
generic value not assigned
```

I'm not really sure what is going on, but a bit of debugging shows that this call does not return an error in the latter case (maybe it should?).

Reply:
Great spot. I have't thought about this case. Will check it out. Thanks!

Reply:
Should be fixed now